In [1]:
# ---
# Imports and Setup
# ---
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display options
pd.set_option("display.max_columns", 100)
sns.set_theme("notebook")

In [3]:
import os
os.getcwd()

'/Users/stefannhs/Projects/open-rec-sphere/notebooks'

In [5]:
# ---
# Imports and Setup
# ---

path = "../data/raw/expedia/"
df_train = pd.read_csv(path + "train.csv")
df_destinations = pd.read_csv(path + "destinations.csv")

## 1. Dataset Familiarization

### Q: What does each row in train.csv represent — a search, a session, or a single interaction?

From the dataset description:

We're looking at logs of customer behavior, which include:
- what customers searched for,
- how they interacted with search results (click/book),
- whether or not the search result was a travel package.

Training data includes all the users in the logs, including both click events and booking events.

Here is the flow that generates the training data:

1. User searches for hotels (with search parameters)
2. Expedia returns a set of hotel options
3. Each hotel option is a row in the training dataset
4. A user can interact with an option (e.g. click, book) - this is tracked

### Q: What is hotel_cluster? Why are there 100 of them?

#### What is hotel_cluster?

`hotel_cluster` is a **categorical label** assigned to each hotel in the dataset, created by **Expedia’s internal clustering algorithms**. These clusters group similar hotels based on:

- Historical booking behavior
- Price levels
- Star ratings
- Geographic proximity (e.g., distance to city center)
- Possibly other latent features

This clustering allows Expedia to **reduce the complexity** of modeling individual hotels and instead focus on groups that behave similarly in terms of user preferences and booking patterns.

#### Why are there 100 of them?

We assume Expedia originally had **many more clusters**, but for the purposes of the Kaggle competition, they:

- **Randomly sampled 100 hotel clusters** from their full set
- Designed the challenge to be a **100-class classification problem**

This constraint makes the problem more tractable for participants, while still simulating a real-world multi-class recommendation scenario.



### Q: Do all columns have intuitive meanings? What about `user_location_city` vs `orig_destination_distance`?

#### General observation

While the column names are fairly descriptive, they are **internally-defined abstractions** rather than human-readable values — so **some semantics are hidden** behind ID mappings (e.g., `user_location_city`, `hotel_market`, `srch_destination_id`). You can interpret many of them through inference and exploratory analysis, but you don’t get explicit labels like "Amsterdam" or "New York".

Expedia did provide a data field description, but **interpretation still requires domain reasoning** — especially for interactions between location, time, and booking behavior.

#### Focus: `user_location_city` vs `orig_destination_distance`

|  Feature |  Description  | Data Type  |
|----------|---------------|------------|
|  `user_location_city`  |  The ID of the city the customer is located  |  int  |
|  `orig_destination_distance`  |  Physical distance between a hotel and a customer at the time of search. A null means the distance could not be calculated  |  double  |

##### Why it matters?
`user_location_city` is a **categorical variable**, and useful for understanding **home-base** or regional trends (e.g. Dutch users booking hotels in Spain during summer).

`orig_destination_distance` is a **continuous numerical variable**, and useful for identifying domestic vs international trips, travel intent, or travel planning behavior.

Their relationship can reveal latent clusters in travel behavior — like whether short-distance searches tend to be booked more often (weekend trips vs intercontinental vacations).




### Q: How many columns in destinations.csv? What's the shape and dtype?

In [22]:
df_destinations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62106 entries, 0 to 62105
Columns: 150 entries, srch_destination_id to d149
dtypes: float64(149), int64(1)
memory usage: 71.1 MB


The destinations.csv file contains:

- **62,106 rows** — each representing a unique `srch_destination_id`
- **150 columns** in total:
    - `srch_destination_id` (type: `int64`)
    - 149 feature columns named d1 to d149 (type: `float64`)
- The dataset uses **~71MB of memory** in its raw form

These 149 $d_{n}$ columns are dense numerical embeddings or latent features that likely capture characteristics of the destination.

Since they’re floats, they could represent:
- Embedding vectors from historical user behavior
- Aggregated signals (e.g., popularity, price profile, geographic features)